handler/src/test/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandlerTest.java

```diff
@@ -0,0 +1,360 @@
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                ctx.writeAndFlush(newOneMessage());
+            }
+
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                //shouldn't be any channel closing, RST, etc from client and no ReadTimeoutExceptions in this test
+                unexpectedExceptionInServerHandler.set(true);
+            }
+        };
+
+        ChannelInboundHandler trackingHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                clientReads.set(true);
+            }
+
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                unexpectedExceptionInClientHandlerLatch.countDown();
+            }
+        };
+
+        Channel server = newServer(pongAfterTwoMessagesHandler);
+        Channel client = newClient(server.localAddress(), readAfterWriteTimeoutHandler, trackingHandler);
+        try {
+            client.writeAndFlush(newOneMessage());
+            assertFalse("Shouldn't get any timeouts or exceptions in client channel",
+                        unexpectedExceptionInClientHandlerLatch.await(2 * TIMEOUT_MS, TimeUnit.MILLISECONDS));
+            assertTrue("Client should get server response already", clientReads.get());
+            assertFalse("Unexpected exception in server handler", unexpectedExceptionInServerHandler.get());
+        } finally {
+            client.close();
+            server.close();
+        }
+    }
+
+    /**
+     * We should continue work after ReadTimeout
+     */
+    @Test
+    public void testContinueCommunicationAfterTimeout() throws Exception {
+        final int TIMEOUT_MS = 300;
+        final int MESSAGES_COUNT = 2;
+
+        final AtomicInteger serverReads = new AtomicInteger();
+        final CountDownLatch timeoutGotLatch = new CountDownLatch(1);
+        final CountDownLatch clientReadsLatch = new CountDownLatch(1);
+        final CountDownLatch suppressedReadsLatch = new CountDownLatch(1);
+        final AtomicBoolean unexpectedExceptionInServerHandler = new AtomicBoolean();
+        final AtomicBoolean unexpectedExceptionInClientHandler = new AtomicBoolean();
+
+        ReadAfterWriteTimeoutHandler readAfterWriteTimeoutHandler =
+                new ReadAfterWriteTimeoutHandler(TIMEOUT_MS, TimeUnit.MILLISECONDS, false, true);
+
+        ChannelInboundHandler pongHandler = new ChannelInboundHandlerAdapter() {
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                serverReads.incrementAndGet();
+                if (serverReads.get() != MESSAGES_COUNT) {
+                    Thread.sleep(2 * TIMEOUT_MS);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2925)